### PR TITLE
Command line fixes and additions, and preserve comments in levels

### LINKED ISSUE #244: Do the render methods need to add all that meta?
Could these go?
```
:num_saved=0
:num_killed=0
:num_waiting=0
:paused=false
```

This roundtrips, and so things get added:
```
./runrabbit text -e ../foo.rel
```
I quite like this for some things (desc, author), as it adds placeholders to remind me. Those 4 bits I mentioned seem extraneous (and also make the tests verbose): are they necessary?

### LINKED ISSUE #248: --interactive mode not working
```
$ make runinteractive 
. Checking for warnings in Makefile
. Checking version number (0.7.1) is consistent everywhere
. Checking for use of java.awt in engine code
. Compiling

BUILD SUCCESSFUL
Total time: 3 seconds
java -cp rabbit-escape-engine/bin/:rabbit-escape-render/bin/:rabbit-escape-ui-text/bin/:rabbit-escape-ui-swing/bin/ rabbitescape.ui.text.TextMain test/level_01.rel --interactive
rabbitescape.engine.util.CommandLineOption$UnknownOption: Unkown option: 'test/level_01.rel'.
	at rabbitescape.engine.util.CommandLineOption.getUnknownOption(CommandLineOption.java:132)
	at rabbitescape.engine.util.CommandLineOptionSet.match(CommandLineOptionSet.java:37)
	at rabbitescape.engine.util.CommandLineOptionSet.parse(CommandLineOptionSet.java:11)
	at rabbitescape.ui.text.TextMain.main(TextMain.java:45)
make: *** [runinteractive] Error 1
```

```
$ java -cp rabbit-escape-engine/bin/:rabbit-escape-render/bin/:rabbit-escape-ui-text/bin/:rabbit-escape-ui-swing/bin/ rabbitescape.ui.text.TextMain --interactive test/level_01.rel
rabbitescape.engine.util.CommandLineOption$UnknownOption: Unkown option: '--interactive'.
	at rabbitescape.engine.util.CommandLineOption.getUnknownOption(CommandLineOption.java:132)
	at rabbitescape.engine.util.CommandLineOptionSet.match(CommandLineOptionSet.java:37)
	at rabbitescape.engine.util.CommandLineOptionSet.parse(CommandLineOptionSet.java:11)
	at rabbitescape.ui.text.TextMain.main(TextMain.java:45)
```

I think I am missing something - @colonelfazackerley how are you generating solutions if you can't run this way?

### PR DESCRIPTION
**Pull request by colonelfazackerley**
…ingle game with text ui

fixes #248 
closes #244

### CODE REVIEW

**Comment by colonelfazackerley** [comment]
I am now trying to get comments to round-trip. if you pull this as-is, and then I push more, can you pull again?

**Comment by colonelfazackerley** [comment]
still need to work on 
* multi-line comment blocks
* comments on ability meta (":block=3" etc)
* comments at the end of the file.

tests for
* comments moving around as round-tripping moves meta etc
* multi line comments

**Comment by colonelfazackerley** [comment]
outstanding
*   comments on ability meta (":block=3" etc)
*   comments at the end of the file.

tests for

 *   comments moving around as round-tripping moves meta etc

**Comment by andybalaam** [comment]
> if you pull this as-is, and then I push more, can you pull again?

I think so, although I would prefer separate pull requests for separate things where possible.

File: rabbit-escape-engine/src/rabbitescape/engine/textworld/Comment.java [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+    }
+
+    public boolean follows( String key )
+    {
+        if ( null == keyFollowing )
+        {
+            return false;
+        }
+        // Compare the starts of the keys only so hint.1 equals hint.1.code etc.
+        int l1 = key.length(), l2 = keyFollowing.length();
```
Comment (andybalaam) [inline rabbit-escape-engine/src/rabbitescape/engine/textworld/Comment.java]: Does this mean the comment for hint.12 will be reported as following hint.1? Would a special case for .code be better?

File: rabbit-escape-engine/src/rabbitescape/engine/textworld/LineProcessor.java [hunk 2]
```diff
@@ -125,6 +127,16 @@ public LineProcessor(
...
+    
+    public Comment[] getComments()
+    {
+        Comment [] ret = new Comment[comments.size()];
+        for ( int i = 0; i < comments.size(); i++ )
+        {
+            ret[i] = comments.get( i );
+        }
+        return ret;
+    }
```
Comment (andybalaam) [inline rabbit-escape-engine/src/rabbitescape/engine/textworld/LineProcessor.java]: Will Collections.toArray do this for us? (Not sure we are allowed to use it in Android API level 8 either.) If not I'm fairly sure I wrote a Util method for this.

File: rabbit-escape-engine/src/rabbitescape/engine/textworld/LineProcessor.java [hunk 3]
```diff
@@ -399,6 +445,7 @@ else if ( value.equals( "false" ) )
 
     private void processItemsLine( String line, VariantGenerator variantGen )
     {
+        maybeLinkToLastComment( Comment.WORLD_ASCII_ART );
         // Treat empty lines as blank lines (Github converts blank lines to empty lines, so it seems sensible to reverse the process).
         if ( line.length() != 0 )
```
Comment (andybalaam) [inline rabbit-escape-engine/src/rabbitescape/engine/textworld/LineProcessor.java]: So comments jump to above the world data. I guess I am ok with this...

Comment (colonelfazackerley) [inline rabbit-escape-engine/src/rabbitescape/engine/textworld/LineProcessor.java]: Interspersing comments within the ASCII art block would be bad. It is good that this would correct that.

File: rabbit-escape-engine/src/rabbitescape/engine/textworld/LineProcessor.java (hunk 2 above)
Comment (colonelfazackerley) [inline rabbit-escape-engine/src/rabbitescape/engine/textworld/LineProcessor.java]: I will look at this.

File: rabbit-escape-engine/src/rabbitescape/engine/textworld/Comment.java (hunk 1 above)
Comment (colonelfazackerley) [inline rabbit-escape-engine/src/rabbitescape/engine/textworld/Comment.java]: good point. i will fix that, and write a test.

File: rabbit-escape-engine/src/rabbitescape/engine/textworld/TextWorldManip.java [hunk 4]
```diff
@@ -210,53 +212,110 @@ public static World createEmptyWorld( int width, int height )
...
+    private static void maybeInsertComment( String key, List<String> ret, Comment[] comments )
+    {
+        for( Comment c: comments)
+        {
+            if( c.follows( key ))
+            {
+                ret.add( c.text );
+            }
+        }
+    }
```
Comment (andybalaam) [inline rabbit-escape-engine/src/rabbitescape/engine/textworld/TextWorldManip.java]: Maybe a map of key to comments would make this feel more natural.

File: rabbit-escape-engine/src/rabbitescape/engine/textworld/LineProcessor.java (hunk 3 above)
Comment (andybalaam) [inline rabbit-escape-engine/src/rabbitescape/engine/textworld/LineProcessor.java]: If it is bad it should be illegal. I'm not convinced it is bad.

**Comment by colonelfazackerley** [comment]
I think my Util.chain with varargs works perfectly now: is there a simpler way?

Is my* Util.concat with varargs going to be OK on Android?

*
Well it's mostly a copy-paste from the internet.

File: rabbit-escape-engine/src/rabbitescape/engine/textworld/TextWorldManip.java [hunk 5]
```diff
@@ -210,53 +212,110 @@ public static World createEmptyWorld( int width, int height )
...
-        ret.add( metaLine( num_rabbits,  world.num_rabbits ) );
-        ret.add( metaLine( num_to_save,  world.num_to_save ) );
-        ret.add( metaLine( rabbit_delay, world.rabbit_delay ) );
-        ret.add( metaLine( num_saved,    world.num_saved ) );
-        ret.add( metaLine( num_killed,   world.num_killed ) );
-        ret.add( metaLine( num_waiting,  world.num_waiting ) );
-        ret.add( metaLine( paused,       world.paused ) );
-
+        maybeInsertComment( name, ret,            world );
+        ret.add( metaLine(  name,             world.name ) );
```
Comment (andybalaam) [inline rabbit-escape-engine/src/rabbitescape/engine/textworld/TextWorldManip.java]: A function for `maybeInsertComment` + `ret.add( metaLine` would make this clearer.

File: rabbit-escape-engine/src/rabbitescape/engine/util/Util.java [hunk 6]
```diff
@@ -420,36 +421,67 @@ public void remove()
...
-                    private final Iterator<? extends T> i1;
-                    private final Iterator<? extends T> i2;
-
-                    public MyIt(
-                        Iterator<? extends T> i1, Iterator<? extends T> i2 )
+                    /** Iterator List */
+                    ArrayList<Iterator<? extends T>> iL;
+                    
+                    /** The position in the ArrayList that we are on */
+                    int indexA = 0;
```
Comment (andybalaam) [inline rabbit-escape-engine/src/rabbitescape/engine/util/Util.java]: Could we use an iterator through iL instead of the index?

File: rabbit-escape-engine/src/rabbitescape/engine/textworld/LineProcessor.java (hunk 2 above)
Comment (colonelfazackerley) [inline rabbit-escape-engine/src/rabbitescape/engine/textworld/LineProcessor.java]: oops, it's just List.toArray

File: rabbit-escape-engine/src/rabbitescape/engine/util/Util.java [hunk 7]
```diff
@@ -631,13 +663,23 @@ protected String nextOrNull()
...
+        @SuppressWarnings( "unchecked" )
+        T[] all = (T[])Array.newInstance(
+            arrays.getClass().getComponentType().getComponentType(), totalLen);
+        int copied = 0;
+        for (T[] arr: arrays)
+        {
+            System.arraycopy(arr, 0, all, copied, arr.length);
+            copied += arr.length;
+        }
+        return all;
```
Comment (andybalaam) [inline rabbit-escape-engine/src/rabbitescape/engine/util/Util.java]: Wouldn't the previous implementation (`list(chain(...`) still work?

File: rabbit-escape-engine/test/rabbitescape/engine/util/TestUtil.java [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+import static org.hamcrest.CoreMatchers.*;
+
+import java.util.Arrays;
+
+import org.junit.Test;
+
+public class TestUtil
+{
+    @Test
+    public void Chain_3a()
```
Comment (andybalaam) [inline rabbit-escape-engine/test/rabbitescape/engine/util/TestUtil.java]: Better name might be Chaining_3_nonempty_arrays_gives_concatenated_list

File: rabbit-escape-engine/test/rabbitescape/engine/util/TestUtil.java [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+        Iterable<String> chained = Util.chain( Arrays.asList( a ), 
+                                               Arrays.asList( _bcd ), 
+                                               Arrays.asList( efg ), 
+                                               Arrays.asList( _h ) );
+        assertThat( Util.list( chained ).toArray( new String[] {} ), 
+            equalTo( new String[] {"a", "e", "f", "g" } ) );
+    }
+    
+    @Test
+    public void Concat_3()
```
Comment (andybalaam) [inline rabbit-escape-engine/test/rabbitescape/engine/util/TestUtil.java]: Better name might be Chain_including_empty_list_skips_it

File: rabbit-escape-engine/test/rabbitescape/engine/util/TestUtil.java [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+        
+        Iterable<String> chained = Util.chain( Arrays.asList( ab ), 
+                                               Arrays.asList( cd ), 
+                                               Arrays.asList( ef ) );
+        assertThat( Util.list( chained ).toArray( new String[] {} ), 
+            equalTo( new String[] {"a", "b", "c", "d", "e", "f" } ) );
+    }
+
+    @Test
+    public void Chain_3b()
```
Comment (andybalaam) [inline rabbit-escape-engine/test/rabbitescape/engine/util/TestUtil.java]: Chaining_3_nonempty_arrays_gives_concatenated_list

File: rabbit-escape-engine/test/rabbitescape/engine/util/TestUtil.java (hunk 9 above)
Comment (andybalaam) [inline rabbit-escape-engine/test/rabbitescape/engine/util/TestUtil.java]: Better name might be Concatenating_3_arrays_gives_all_elements

File: rabbit-escape-ui-text/src/rabbitescape/ui/text/TextMain.java [hunk 11]
```diff
@@ -29,31 +28,43 @@ public TextMain( FileSystem fs, Terminal terminal, Config config )
...
+                                        level, solution, encode, decode, help, 
+                                        noinput, placeholders, template );
             if ( help.isPresent() )
             {
                 usageMessage();
                 System.exit( 0 );
             }
+            if ( noinput.isPresent() )
+            {
+                TextSingleGameEntryPoint.entryPoint( new String[] {noinput.getValue(), "noinput"} );
```
Comment (andybalaam) [inline rabbit-escape-ui-text/src/rabbitescape/ui/text/TextMain.java]: Sorry for the mis-lead, but is there any value in `--noinput` at all any more?

File: rabbit-escape-engine/src/rabbitescape/engine/textworld/LineProcessor.java (hunk 2 above)
Comment (andybalaam) [inline rabbit-escape-engine/src/rabbitescape/engine/textworld/LineProcessor.java]: Cool :-)

**Comment by andybalaam** [comment]
> Well it's mostly a copy-paste from the internet.

You'd better confirm the license of what you copied, or state here that you took the ideas and re-wrote the code...

**Comment by andybalaam** [comment]
> I think my Util.chain with varargs works perfectly now: is there a simpler way?

It looks reasonable to me.  I'd like to see test cases for:

- zero and 1-length arrays
- chaining arrays of things of different types.

File: rabbit-escape-engine/src/rabbitescape/engine/util/Util.java (hunk 7 above)
Comment (andybalaam) [inline rabbit-escape-engine/src/rabbitescape/engine/util/Util.java]: If not, we need to revisit this - casting the result of Array.newInstance is evil, and possibly fails in some cases.  It's evil enough for me to want to kill it even if it works.

**Comment by andybalaam** [comment]
Want me to merge some or all of this?  I could probably merge the CLI stuff, leaving us a simpler diff to discuss for the other stuff.

File: rabbit-escape-engine/src/rabbitescape/engine/util/Util.java (hunk 7 above)
Comment (colonelfazackerley) [inline rabbit-escape-engine/src/rabbitescape/engine/util/Util.java]: No. Runtime error from toArray(arrays[0])

**Comment by colonelfazackerley** [comment]
> You'd better confirm the license of what you copied, or state here that you took the ideas and re-wrote the code...

http://stackoverflow.com/questions/9476613/simplify-this-generic-method-to-concatenate-java-arrays
from user newacct
http://stackoverflow.com/users/86989/newacct
CC-SA 3
http://creativecommons.org/licenses/by-sa/3.0/
http://blog.stackoverflow.com/2009/06/attribution-required/

So, it's OK, but the links above are needed. I can change the code, if you don't like this.

**Comment by colonelfazackerley** [comment]
I guess comments in the code next to the method are OK for the links. It is on the web via github.

**Comment by colonelfazackerley** [comment]
> Want me to merge some or all of this? I could probably merge the CLI stuff, leaving us a simpler diff to discuss for the other stuff.

OK.

then I can finish the comment stuff, and you could merge that. 

Then I can finish the Util stuff.

File: rabbit-escape-engine/src/rabbitescape/engine/util/Util.java (hunk 7 above)
Comment (colonelfazackerley) [inline rabbit-escape-engine/src/rabbitescape/engine/util/Util.java]: OK. I will have another go at this. I understand the objection to the array cast.

This also removes the CCSA attribution pain.

I was wondering about calling the 2 arg concat recursively to do varargs.

**Comment by andybalaam** [comment]
My reading of https://www.gnu.org/licenses/license-list.html#ccbysa is that CC BY-SA can't be incorporated into a GPLv2 project (only GPLv3) so, annoying as it is, please can you rewrite this snippet I your own way? It can use the ideas, but not be a direct copy. Sorry for the hassle.

**Comment by colonelfazackerley** [comment]
TODO 
- iterator for iL in Util.chain
- rewrite Util.concat
- rename tests for chain and concat
- more tests for chain (zero and 1-length arrays, chaining arrays of things of different types.)

i think the comment stuff is done. you can merge that now, or after i rebase. (an idea of how much rebasing to do would be useful for me)

**Comment by colonelfazackerley** [comment]
That's a temporary replacement for concat. If I have a brainwave, I will have another go at a proper varargs implementation.

**Comment by colonelfazackerley** [comment]
> Sorry for the mis-lead, but is there any value in ```--noinput``` at all any more?

I don't have a use for it. We do have ```--level foo.rel --solution 1```, which I do find useful.

Shall I remove  ```--noinput```?

**Comment by andybalaam** [comment]
> Shall I remove --noinput?

Yes, but I am losing track of this pull request, so not here if you can bear with me.

**Comment by colonelfazackerley** [comment]
I think I have covered all comments. This is ready to pull.

**Comment by andybalaam** [comment]
For the record, the code copied from StackOverflow has now been completely re-written.

**Comment by andybalaam** [comment]
I have now rebased and merged this entire pull request.

**Comment by colonelfazackerley** [comment]
Thanks for doing the rebasing. I could have done that.

Probably less painful to do it yourself.

**Comment by andybalaam** [comment]
Rebasing was pretty easy, but np :-). I am a huge fan of `git rebase -i --exec 'make clean slowtest' master`
